### ISSUE TITLE
Update Node.js version support (18, 20, 22)

### ISSUE THREAD

**Issue by orangejulius** [pelias/pelias#950]
It's been a while since we've performed the regular chore of adding support for newer Node.js versions.

The last one we did was [Node.js 12 back in early 2020](https://github.com/pelias/pelias/issues/800), and v12 is now EOL.

I propose we go straight to the newest versions of Node.js, adding support for all currently maintained versions: 16, 18, and 20.

## What needs to be done

For each of our [code repositories](https://github.com/pelias/documentation/blob/master/development/repository_list.md), we need to:
- fix anything in our _own_ code that does not support these new versions,
- Update any dependencies with the same issue and then:
- Update the versions of Node.js used in CI
- Update the `Dockerfile`s to use a base image with a newer Node.js version

## The usual suspects

Generally we've found Node.js upgrades aren't too hard, with most of the work being repetitive changes to CI and Docker configs.

The things that usually cause issue are:
 - the better-sqlite3 Node.js module, which we use extensively
 - the interpolation service, and specifically it's dependency on libpostal through the [node-postal](https://github.com/openvenues/node-postal/) npm module

## Please link to this issue

Please link to this tracking issue in any commits or PR that involve these Node.js upgrades, so we can find the status of everything in one place.

**Comment by orangejulius** [comment]
Alright, I did just a little bit of investigation into this and here's what I've found:
- "Node.js only" repos like the API will [support up to Node.js 20 just fine](https://github.com/pelias/api/actions/runs/5439357002)
- `better-sqlite3` version 8.4.0 (we use 7.x everywhere right now) is required for Node.js support
- `node-postal` currently works with Node.js up to 16
- It looks like Nodejs 18 and 20 require software packages included in Ubuntu 20, which we use in CI, but are not included in Ubuntu 18, which we currently base all our Docker images on

Based on that last one, we might want to upgrade everything to Node.js 16 for now, just to get off an EOL Node.js version, then upgrade our Docker images to Ubuntu 20 (which might require a bit of work itself), and then work on Node.js 18 and 20 support. We also might want to support different versions of Node.js temporarily for the interpolation service/node-postal, so that all the other packages can use a newer version sooner.

**Comment by orangejulius** [comment]
We're once again behind with this, I plan on making some sweeping changes for consistency. I've updated the title of this issue to reflect the Node.js versions we currently want to support.

1. Update the Docker base image to Node.js 18. All our Docker images will build on Node.js 18 thankfully, though already some (like interpolation) don't build on Node.js 16
2. Update the Node.js versions in CI to include only Node.js 18 and later (12, 14, and 16 are common in our repos but are all EOL). Most repos will currently support Node.js 20 and 22, but some may not yet, if they have dependencies on packages that compile code like `node-postal` (I don't even know what versions `node-postal` currently supports), or `better-sqlite3`
3. Update any Github Actions like `setup-node` and `checkout` so that they are at the latest version. Older versions may run on EOL versions of Node.js and trigger warnings in Github Actions. This is a bit more minor but we might as well.

**Comment by orangejulius** [comment]
Okay, we all know I love a good hacky bash script that is iterated on so much it actually becomes somewhat polished.

Over the past couple days I've borrowed some time here and there to see where our CI falls short or is inconsistent across basically all of our repositories. And I've created this massive script to do up to 6 things per repository, depending on what's needed:

1. update any setup-node or checkout github actions from @v2 or @v2-beta actions to @v4 to avoid any deprecation warnings in CI                                                                                                                
2. change any hardcoded ubuntu 16/18/20 versions to 22 to ensure we're consistent there (see https://github.com/pelias/pelias/issues/950)                                                                                                                                            
3. change ubuntu definition from variable to hardcoded (the variable breaks 3rd party contributors)                                                                                                       4. remove Node.js EOL node versions like 12 (yes we had it in some repos), 14, 16 from CI list, ensure active Node.js versions are in the list if the repository currently works with it (most repos support up to Node.js 22, a few up to 20, and one just 18)
5. normalize github actions workflow files from templates                                                                                                                                                 
6. upgrade docker to use new baseimage (with a no-op feat: BREAKING CHANGE commit)

The script pushes to a new branch and it's currently passing CI across _all repositories. 🎉. Over the coming days I will create the actual PRs and get everything upgraded.

Note that some repos were so far behind CI wouldn't even pass any more for _any_ new branch. So it's fairly important to get these PRs merged to unblock future work. There will be some immediate followup like NPM dependency upgrades and I think we can get to supporting Node.js 22 everywhere pretty soon.

<details>
  <summary>Here's the script, shield your eyes. Run it in a local directory with all pelias repos checked out</summary>

  
  ```bash
#!/bin/bash

set -euo pipefail
set -x

# things to do:
# 1. update any setup-node or checkout github actions from @v2 or @v2-beta actions to @v4
# 2. change any hardcoded ubuntu 16/18/20 versions to 22
# 3. change ubuntu definition from variable to hardcoded (the variable breaks 3rd party contributors)
# 4. remove Node.js 16 from CI list, ensure Node.js 18/20 are active (22 will come later after some upgrades)
# 5. normalize github actions workflow files from templates
# 6. upgrade docker to use new baseimage (with a no-op feat: BREAKING CHANGE commit)


# categorize all our repositories by various factors
importers="openstreetmap whosonfirst openaddresses polylines geonames transit csv-importer"
services="api placeholder libpostal-service interpolation pip-service"
baseimage="baseimage"
libpostal_baseimage="docker-libpostal_baseimage"
schema="schema"
modules="query pbf2json parser blacklist-stream logger dbclient config sorting microservice-wrapper mock-logger labels wof-admin-lookup fuzzy-tester model polygon-lookup analysis"

# don't overwrite the workflow files with template because we do something special
repos_with_custom_workflow_files="interpolation schema pbf2json libpostal-service"
# repos where we should not run precommit hooks since tests fail and the commits would not be made
repos_with_failing_tests="pbf2json" # https://github.com/pelias/pbf2json/issues/108

# some repositories are not ready for the latest Node.js version so record those here
limit_to_node_18="interpolation" # node-postal does not work on node 20 or 22 yet
limit_to_node_20="placeholder pip-service wof-admin-lookup parser $importers" # anything depending on better-sqlite3 will have to be upgraded later

# these are the repos to actually update
repos="schema whosonfirst"

#git_options="--no-verify" # use this line for faster iteration
git_options="" # use this line for final release

# for repos with failing tests/precommit hooks we skip the pre-commit hooks so the commits work
git_verify="git commit"
git_no_verify="git commit --no-verify"

# now create a _bunch_ of template files in variables for the 3 types of github workflow files:
# _test.yml, which stores common stuff for unit tests, etc
# pull_request.yml, which manages what should run for _external_ contributions
# push.yml, which manages what should run for the master branch and internal contributions
#
# note: there are two variations of the push workflow to account for repos without a dockerfile
github_workflow_file_test=$(cat <<'EOF'
name: Unit Tests
on: workflow_call
jobs:
  unit-tests:
    runs-on: '${{ matrix.os }}'
    strategy:
      matrix:
        os:
          - ubuntu-22.04
        node-version: [ 18.x, 20.x, 22.x ]
    steps:
      - uses: actions/checkout@v4
      - name: 'Install node.js ${{ matrix.node-version }}'
        uses: actions/setup-node@v4
        with:
          node-version: '${{ matrix.node-version }}'
      - name: Run unit tests
        run: |
          [[ -f ./bin/ci-setup ]] && ./bin/ci-setup
          npm install
          npm run ci
EOF
)

github_workflow_file_push=$(cat <<'EOF'
name: Continuous Integration
on: push
jobs:
  unit-tests:
    uses: ./.github/workflows/_test.yml
  npm-publish:
    needs: unit-tests
    if: github.ref == 'refs/heads/master' && needs.unit-tests.result == 'success'
    runs-on: ubuntu-22.04
    steps:
      - uses: actions/checkout@v4
      - name: Install Node.js
        uses: actions/setup-node@v4
        with:
          node-version: 20.x
      - name: Run semantic-release
        env:
          GH_TOKEN: ${{ secrets.GH_SEMANTIC_RELEASE_TOKEN }}
          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
        run: >
          if [[ -n "$GH_TOKEN" && -n "$NPM_TOKEN" ]]; then
            curl "https://raw.githubusercontent.com/pelias/ci-tools/master/semantic-release.sh" | bash -
          fi
  build-docker-images:
    # run this job if the unit tests passed and the npm-publish job was a success or was skipped
    # note: github actions won't run a job if you don't call one of the status check functions, so `always()` is called since it evalutes to `true`
    if: ${{ always() && needs.unit-tests.result == 'success' && (needs.npm-publish.result == 'success' || needs.npm-publish.result == 'skipped') }}
    needs: [unit-tests, npm-publish]
    runs-on: ubuntu-22.04
    steps:
      - uses: actions/checkout@v4
      - name: Build Docker images
        env:
          DOCKER_USERNAME: ${{ secrets.DOCKER_USERNAME }}
          DOCKER_PASSWORD: ${{ secrets.DOCKER_PASSWORD }}
        run: |
          curl "https://raw.githubusercontent.com/pelias/ci-tools/master/build-docker-images.sh" | bash -
EOF
)

github_workflow_file_push_no_dockerfile=$(cat <<'EOF'
name: Continuous Integration
on: push
jobs:
  unit-tests:
    uses: ./.github/workflows/_test.yml
  npm-publish:
    needs: unit-tests
    if: github.ref == 'refs/heads/master' && needs.unit-tests.result == 'success'
    runs-on: ubuntu-22.04
    steps:
      - uses: actions/checkout@v4
      - name: Install Node.js
        uses: actions/setup-node@v4
        with:
          node-version: 20.x
      - name: Run semantic-release
        env:
          GH_TOKEN: ${{ secrets.GH_SEMANTIC_RELEASE_TOKEN }}
          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
        run: >
          if [[ -n "$GH_TOKEN" && -n "$NPM_TOKEN" ]]; then
            curl "https://raw.githubusercontent.com/pelias/ci-tools/master/semantic-release.sh" | bash -
          fi
EOF
)

github_workflow_file_pull_request=$(cat <<'EOF'
name: Continuous Integration
on: pull_request
jobs:
  unit-tests:
    # only run this job for forks
    if: github.event.pull_request.head.repo.full_name != github.repository
    uses: ./.github/workflows/_test.yml
EOF
)

echo $repos

for repo in $repos; do
	echo $repo
	pushd $repo
	pwd

	[[ $repos_with_failing_tests == *$repo* ]] && git_command=$git_no_verify || git_command=$git_verify

	# force the repo into a good state on the master branch before creating a working branch
	git stash
	git checkout master
	git fetch
	git pull
	git cleanup || true

	branch_name="nodejs-18-and-ci-updates"
	git branch -D $branch_name || true
	git checkout -b $branch_name || true
	git checkout $branch_name

	if [[ -f package.json ]]; then
		rm -rf node_modules
		npm install
	fi

	# 1. update checkout and setup-node versions
	sed -i '' -E 's/checkout\@v2(-beta)?/checkout@v4/g' .github/workflows/*.yml
	sed -i '' -E 's/setup-node\@v2(-beta)?/setup-node@v4/g' .github/workflows/*.yml

	git add .github
	git diff --staged
	$git_command $git_options -m "CI: update checkout and setup-node actions to latest

This just keeps us up to date and avoids warnings for deprecated Node.js versions in CI." || true

	# 2 update hardcoded ubuntu versions to 20
	sed -i '' -E 's/ubuntu-(16|18|20).04/ubuntu-22.04/g' .github/workflows/*.yml

	git add .github
	git diff --staged
	$git_command $git_options -m "CI: update all hardcoded Ubuntu versions to 22.04

We were all over the place with these, with some going all the way back to Ubuntu 16!

Connects https://github.com/pelias/pelias/issues/951" || true

	# 3 update ubuntu versions from GH Actions variable back to hardcoded
	sed -i '' 's/\${{ vars\.UBUNTU_VERSION }}/ubuntu-22.04/g' .github/workflows/*.yml

	git add .github
	git diff --staged
	$git_command $git_options -m "CI: move away from an Ubuntu version variable in GH Actions

This doesn't really save us much effort and breaks CI on forks.

Connects https://github.com/pelias/pelias/issues/951" || true

	# 4. update Node.js versions
	sed -i '' -E '/- (12|14|16)\.x/d' .github/workflows/*.yml

	git add .github
	git diff --staged
	$git_command $git_options -m "CI: remove EOL Node.js 12/16/18" -m "https://github.com/pelias/pelias/issues/950" || true

	# replace multi-line node version specification with one line that will be easier to grep in the future
	sed -i '' -E '/- (18|20|22)\.x/d' .github/workflows/*.yml

	# not all repos support up to Node.js 22 yet, so update according to an exception list
	if [[ $limit_to_node_18 == *$repo* ]]; then
		sed -i '' 's/node-version:$/node-version: [ 18.x ]/' .github/workflows/*.yml
		message1="CI: Test against Node.js version 18"
		message2="We will have to upgrade node-postal to 1.20 to support Node.js 20, and better-sqlite3 before we can support Node.js 22.

This also rewrites our CI config so that all Node.js versions are on one line for ease of future grepping.

Connects https://github.com/pelias/pelias/issues/950" || true
	elif [[ $limit_to_node_20 == *$repo* ]]; then
		sed -i '' 's/node-version:$/node-version: [ 18.x, 20.x ]/' .github/workflows/*.yml
		message1="CI: Test against Node.js versions 18 and 20"
		message2="We will have to upgrade better-sqlite3 before we can support Node.js 22

This also rewrites our CI config so that all Node.js versions are on one line for ease of future grepping.

Connects https://github.com/pelias/pelias/issues/950" || true
	else
		sed -i '' 's/node-version:$/node-version: [ 18.x, 20.x, 22.x ]/' .github/workflows/*.yml
		message1="CI: Test against Node.js versions 18 ,20, and 22"
		message2="This means we support all currently supported LTS releases of Node.js!

Also, rewrite our CI config so that all Node.js versions are on one line for ease of future grepping.

Connects https://github.com/pelias/pelias/issues/950" || true
	fi

	git add .github
	git diff --staged
	$git_command $git_options -m "$message1" -m "$message2" || true

	# 5. ensure correct github action file layout
	if [[ ! "$repos_with_custom_workflow_files" == *$repo* ]]; then
		echo "repo should use new format of workflow files"

		echo "$github_workflow_file_test" > .github/workflows/_test.yml
		if [[ -f Dockerfile ]]; then
			echo "$github_workflow_file_push" > .github/workflows/push.yml
		else
			echo "$github_workflow_file_push_no_dockerfile" > .github/workflows/push.yml
		fi

		echo "$github_workflow_file_pull_request" > .github/workflows/pull_request.yml

		# the templates use node 18-22, so adjust again based on this repos needs
		if [[ $limit_to_node_18 == *$repo* ]]; then
			sed -i '' 's/node-version: \[.*$/node-version: [ 18.x ]/' .github/workflows/_test.yml
		elif [[ $limit_to_node_20 == *$repo* ]]; then
			sed -i '' 's/node-version: \[.*$/node-version: [ 18.x, 20.x ]/' .github/workflows/_test.yml
		fi

		git add .github
		$git_command $git_options -m "CI: use standard Github Action templates" -m "A lot of our repositories have diverged from our intended template, so this copies a templated version over whatever was here before

Connects https://github.com/pelias/pelias/issues/951" || true
	fi

	# 6. For repos with Dockerfile, create a breaing change no-op commit announcing the change to a baseimage with Node.js 18
	if [[ -f Dockerfile ]]; then
		$git_command $git_options --allow-empty -m "feat(docker): Upgrade to Node.js 18 Docker baseimage" -m "BREAKING CHANGE: The Docker baseimage has been updated to use Node.js 18.20.5, so this repository's Docker image will also use it.

https://github.com/pelias/pelias/issues/950" || true
	fi

	git push --force-with-lease
	popd
done
  ```
  
</details>

**Comment by orangejulius** [comment]
Okay, all or repositories now support Node.js 18, 20, and 22 _except_ the interpolation repository. That one is waiting for a node-postal release with https://github.com/openvenues/node-postal/pull/51 included. Then we're all done!

**Comment by orangejulius** [comment]
We Pelias maintainers now have publish access to [node-postal](github.com/openvenues/node-postal), and we've published a new version that supports all modern Node.js versions. With that the last few Pelias repos that don't support Node.js 22 are updated.

Which is great because it's about time to drop Node.js 18 and add support for Node.js 24 😆
